### PR TITLE
Gh 1134 Serialiser Consistency

### PR DESCRIPTION
**Pull request by m607123**
Simple method implemented to (subjectively) indicate whether a serialiser is consistent.
Added a consistency check into the schema validation, within the Store abstract class - if a serialiser is not consistent, it will add an error to the ValidationResult, stating which serialiser has been selected, as well as informing the user that it is inconsistent and the store may be unable to handle it.

### CODE REVIEW

**Comment by m607123** [comment]
Closing for now - work still to be done.

File: store-implementation/hbase-store/src/main/java/uk/gov/gchq/gaffer/hbasestore/HBaseStore.java [hunk 1]
```diff
@@ -180,6 +184,25 @@ protected SchemaOptimiser createSchemaOptimiser() {
...
+        }
+    }
+
+    @Override
+    protected void validateSchemaElementDefinition(final Entry<String, SchemaElementDefinition> schemaElementDefinitionEntry, final ValidationResult validationResult) {
+        super.validateSchemaElementDefinition(schemaElementDefinitionEntry, validationResult);
+
+        for (String property : schemaElementDefinitionEntry.getValue().getGroupBy()) {
+            if (!schemaElementDefinitionEntry.getValue().getPropertyTypeDef(property).getSerialiser().isConsistent()) {
+                validationResult.addError("Property serialiser is inconsistent - store may be unable to handle this.");
```
Comment (p013570) [inline store-implementation/hbase-store/src/main/java/uk/gov/gchq/gaffer/hbasestore/HBaseStore.java]: Please could you update the error message to:

"Serialiser for groupBy property " + property + " is inconsistent. This store requires all groupBy property serialisers to be consistent."

File: store-implementation/hbase-store/src/main/java/uk/gov/gchq/gaffer/hbasestore/HBaseStore.java [hunk 2]
```diff
@@ -180,6 +184,25 @@ protected SchemaOptimiser createSchemaOptimiser() {
     }
 
     @Override
+    public void validateSchemas() {
+        super.validateSchemas();
+        if (!getSchema().getVertexSerialiser().isConsistent()) {
+            throw new SchemaException("Vertex serialiser is inconsistent - store may be unable to handle this.");
```
Comment (p013570) [inline store-implementation/hbase-store/src/main/java/uk/gov/gchq/gaffer/hbasestore/HBaseStore.java]: Please could you update the error message to:

"Vertex serialiser is inconsistent. This store requires vertices to be serialised in a consistent way."

File: store-implementation/hbase-store/src/main/java/uk/gov/gchq/gaffer/hbasestore/HBaseStore.java [hunk 3]
```diff
@@ -180,6 +184,25 @@ protected SchemaOptimiser createSchemaOptimiser() {
     }
 
     @Override
+    public void validateSchemas() {
```
Comment (p013570) [inline store-implementation/hbase-store/src/main/java/uk/gov/gchq/gaffer/hbasestore/HBaseStore.java]: So it looks like Accumulo, HBase and Parquet stores all have this same validation. Perhaps you could create 2 methods in Store with this logic.

validateConsistentVertex and validateConsistentGroupByProperties.

Then in the store implementations we can just call those 2 methods.

**Comment by p013570** [comment]
I've just noticed that the SchemaOptimiser.addDefaultSerialisers and SchemaOptimiser.getDefaultVertexSerialiser also need to be updated. We should pick serialisers that are consistent for the vertex serialiser and groupBy properties.

Unfortunately this will mean a potential migration step for users. Any users that have existing data need to manually hardcode all the serialisers they are currently using in their schema to avoid different ones being used. Similar to the migration steps for this issue: https://github.com/gchq/Gaffer/issues/988

**Comment by m607123** [comment]
Closing while further changes made.

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/serialiser/lengthvalue/EdgeIdSerialiser.java [hunk 4]
```diff
@@ -89,4 +89,9 @@ public EdgeId deserialiseEmpty() throws SerialisationException {
     public boolean preservesObjectOrdering() {
         return false;
     }
+
+    @Override
+    public boolean isConsistent() {
+        return true;
```
Comment (p013570) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/serialiser/lengthvalue/EdgeIdSerialiser.java]: This should actually be:
return null != vertexSerialiser && vertexSerialiser.isConsistent();

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/serialiser/lengthvalue/EdgeIdSerialiser.java [hunk 5]
```diff
@@ -89,4 +89,9 @@ public EdgeId deserialiseEmpty() throws SerialisationException {
     public boolean preservesObjectOrdering() {
         return false;
```
Comment (p013570) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/serialiser/lengthvalue/EdgeIdSerialiser.java]: Please could you also update this to:

return null != vertexSerialiser && vertexSerialiser.preservesObjectOrdering();

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/serialiser/lengthvalue/EdgeSerialiser.java [hunk 6]
```diff
@@ -100,4 +100,9 @@ public Edge deserialise(final byte[] bytes) throws SerialisationException {
     public Edge deserialiseEmpty() throws SerialisationException {
         return null;
     }
+
+    @Override
+    public boolean isConsistent() {
+        return true;
```
Comment (p013570) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/serialiser/lengthvalue/EdgeSerialiser.java]: this should be false

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/serialiser/lengthvalue/ElementIdSerialiser.java [hunk 7]
```diff
@@ -103,4 +103,9 @@ public ElementId deserialiseEmpty() throws SerialisationException {
     public boolean preservesObjectOrdering() {
         return false;
     }
+
+    @Override
+    public boolean isConsistent() {
+        return true;
```
Comment (p013570) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/serialiser/lengthvalue/ElementIdSerialiser.java]: This should actually be:
return null != vertexSerialiser && vertexSerialiser.isConsistent();

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/serialiser/lengthvalue/ElementIdSerialiser.java [hunk 8]
```diff
@@ -103,4 +103,9 @@ public ElementId deserialiseEmpty() throws SerialisationException {
     public boolean preservesObjectOrdering() {
         return false;
```
Comment (p013570) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/serialiser/lengthvalue/ElementIdSerialiser.java]: Please could you change this to:
return null != vertexSerialiser && vertexSerialiser.preservesObjectOrdering();

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/serialiser/lengthvalue/ElementSerialiser.java [hunk 9]
```diff
@@ -76,4 +76,9 @@ public String getGroup(final byte[] bytes) throws SerialisationException {
     public Element deserialiseEmpty() throws SerialisationException {
         return null;
     }
+
+    @Override
+    public boolean isConsistent() {
+        return true;
```
Comment (p013570) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/serialiser/lengthvalue/ElementSerialiser.java]: this should be false

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/serialiser/lengthvalue/EntityIdSerialiser.java [hunk 10]
```diff
@@ -78,4 +78,9 @@ public EntityId deserialiseEmpty() throws SerialisationException {
     public boolean preservesObjectOrdering() {
         return false;
```
Comment (p013570) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/serialiser/lengthvalue/EntityIdSerialiser.java]: Please could you change this to:

return null != vertexSerialiser && vertexSerialiser.preservesObjectOrdering();

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/serialiser/lengthvalue/EntitySerialiser.java [hunk 11]
```diff
@@ -94,4 +94,9 @@ public Entity deserialise(final byte[] bytes) throws SerialisationException {
     public Entity deserialiseEmpty() throws SerialisationException {
         return null;
     }
+
+    @Override
+    public boolean isConsistent() {
+        return true;
```
Comment (p013570) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/serialiser/lengthvalue/EntitySerialiser.java]: This should be false.

File: core/type/src/main/java/uk/gov/gchq/gaffer/serialisation/AvroSerialiser.java [hunk 12]
```diff
@@ -107,4 +107,9 @@ private void close(final Closeable close) {
             }
         }
     }
+
+    @Override
+    public boolean isConsistent() {
+        return true;
```
Comment (p013570) [inline core/type/src/main/java/uk/gov/gchq/gaffer/serialisation/AvroSerialiser.java]: This should be false.

File: library/bitmap-library/src/main/java/uk/gov/gchq/gaffer/bitmap/serialisation/RoaringBitmapSerialiser.java [hunk 13]
```diff
@@ -71,6 +71,11 @@ public boolean preservesObjectOrdering() {
     }
 
     @Override
+    public boolean isConsistent() {
+        return true;
```
Comment (p013570) [inline library/bitmap-library/src/main/java/uk/gov/gchq/gaffer/bitmap/serialisation/RoaringBitmapSerialiser.java]: This should be false.

File: library/sketches-library/src/main/java/uk/gov/gchq/gaffer/sketches/clearspring/cardinality/serialisation/HyperLogLogPlusSerialiser.java [hunk 14]
```diff
@@ -55,4 +55,9 @@ public HyperLogLogPlus deserialiseEmpty() throws SerialisationException {
     public boolean preservesObjectOrdering() {
         return false;
     }
+
+    @Override
+    public boolean isConsistent() {
+        return true;
```
Comment (p013570) [inline library/sketches-library/src/main/java/uk/gov/gchq/gaffer/sketches/clearspring/cardinality/serialisation/HyperLogLogPlusSerialiser.java]: This should be false.

File: library/sketches-library/src/main/java/uk/gov/gchq/gaffer/sketches/datasketches/cardinality/serialisation/HllUnionSerialiser.java [hunk 15]
```diff
@@ -50,4 +50,9 @@ public Union deserialiseEmpty() throws SerialisationException {
     public boolean preservesObjectOrdering() {
         return false;
     }
+
+    @Override
+    public boolean isConsistent() {
+        return true;
```
Comment (p013570) [inline library/sketches-library/src/main/java/uk/gov/gchq/gaffer/sketches/datasketches/cardinality/serialisation/HllUnionSerialiser.java]: This should be false.

File: library/sketches-library/src/main/java/uk/gov/gchq/gaffer/sketches/datasketches/cardinality/serialisation/HllSketchSerialiser.java [hunk 16]
```diff
@@ -50,4 +50,9 @@ public HllSketch deserialiseEmpty() throws SerialisationException {
     public boolean preservesObjectOrdering() {
         return false;
     }
+
+    @Override
+    public boolean isConsistent() {
+        return true;
```
Comment (p013570) [inline library/sketches-library/src/main/java/uk/gov/gchq/gaffer/sketches/datasketches/cardinality/serialisation/HllSketchSerialiser.java]: This should be false.

File: library/sketches-library/src/main/java/uk/gov/gchq/gaffer/sketches/datasketches/quantiles/serialisation/DoublesSketchSerialiser.java [hunk 17]
```diff
@@ -51,5 +51,10 @@ public DoublesSketch deserialiseEmpty() throws SerialisationException {
     public boolean preservesObjectOrdering() {
         return false;
     }
+
+    @Override
+    public boolean isConsistent() {
+        return true;
```
Comment (p013570) [inline library/sketches-library/src/main/java/uk/gov/gchq/gaffer/sketches/datasketches/quantiles/serialisation/DoublesSketchSerialiser.java]: This should probably be false.

File: library/sketches-library/src/main/java/uk/gov/gchq/gaffer/sketches/datasketches/quantiles/serialisation/DoublesUnionSerialiser.java [hunk 18]
```diff
@@ -53,5 +53,10 @@ public DoublesUnion deserialiseEmpty() throws SerialisationException {
     public boolean preservesObjectOrdering() {
         return false;
     }
+
+    @Override
+    public boolean isConsistent() {
+        return true;
```
Comment (p013570) [inline library/sketches-library/src/main/java/uk/gov/gchq/gaffer/sketches/datasketches/quantiles/serialisation/DoublesUnionSerialiser.java]: This should probably be false.

File: library/sketches-library/src/main/java/uk/gov/gchq/gaffer/sketches/datasketches/quantiles/serialisation/StringsSketchSerialiser.java [hunk 19]
```diff
@@ -57,4 +57,9 @@ public boolean canHandle(final Class clazz) {
     public boolean preservesObjectOrdering() {
         return false;
     }
+
+    @Override
+    public boolean isConsistent() {
+        return true;
```
Comment (p013570) [inline library/sketches-library/src/main/java/uk/gov/gchq/gaffer/sketches/datasketches/quantiles/serialisation/StringsSketchSerialiser.java]: This should probably be false.

File: library/sketches-library/src/main/java/uk/gov/gchq/gaffer/sketches/datasketches/quantiles/serialisation/StringsUnionSerialiser.java [hunk 20]
```diff
@@ -57,4 +57,9 @@ public boolean canHandle(final Class clazz) {
     public boolean preservesObjectOrdering() {
         return false;
     }
+
+    @Override
+    public boolean isConsistent() {
+        return true;
```
Comment (p013570) [inline library/sketches-library/src/main/java/uk/gov/gchq/gaffer/sketches/datasketches/quantiles/serialisation/StringsUnionSerialiser.java]: This should probably be false.

File: library/sketches-library/src/main/java/uk/gov/gchq/gaffer/sketches/datasketches/sampling/serialisation/ReservoirItemsSketchSerialiser.java [hunk 21]
```diff
@@ -56,4 +56,9 @@ public boolean canHandle(final Class clazz) {
     public boolean preservesObjectOrdering() {
         return false;
     }
+
+    @Override
+    public boolean isConsistent() {
+        return true;
```
Comment (p013570) [inline library/sketches-library/src/main/java/uk/gov/gchq/gaffer/sketches/datasketches/sampling/serialisation/ReservoirItemsSketchSerialiser.java]: This should probably be false.

File: library/sketches-library/src/main/java/uk/gov/gchq/gaffer/sketches/datasketches/sampling/serialisation/ReservoirLongsSketchSerialiser.java [hunk 22]
```diff
@@ -51,5 +51,10 @@ public ReservoirLongsSketch deserialiseEmpty() throws SerialisationException {
     public boolean preservesObjectOrdering() {
         return false;
     }
+
+    @Override
+    public boolean isConsistent() {
+        return true;
+    }
 }
```
Comment (p013570) [inline library/sketches-library/src/main/java/uk/gov/gchq/gaffer/sketches/datasketches/sampling/serialisation/ReservoirLongsSketchSerialiser.java]: This should probably be false.

File: library/sketches-library/src/main/java/uk/gov/gchq/gaffer/sketches/datasketches/sampling/serialisation/ReservoirLongsUnionSerialiser.java [hunk 23]
```diff
@@ -51,5 +51,10 @@ public ReservoirLongsUnion deserialiseEmpty() throws SerialisationException {
     public boolean preservesObjectOrdering() {
         return false;
     }
+
+    @Override
+    public boolean isConsistent() {
+        return true;
```
Comment (p013570) [inline library/sketches-library/src/main/java/uk/gov/gchq/gaffer/sketches/datasketches/sampling/serialisation/ReservoirLongsUnionSerialiser.java]: This should probably be false.

File: library/sketches-library/src/main/java/uk/gov/gchq/gaffer/sketches/datasketches/sampling/serialisation/ReservoirNumbersUnionSerialiser.java [hunk 24]
```diff
@@ -53,4 +53,9 @@ public boolean canHandle(final Class clazz) {
     public boolean preservesObjectOrdering() {
         return false;
     }
+
+    @Override
+    public boolean isConsistent() {
+        return true;
```
Comment (p013570) [inline library/sketches-library/src/main/java/uk/gov/gchq/gaffer/sketches/datasketches/sampling/serialisation/ReservoirNumbersUnionSerialiser.java]: This should probably be false.

File: library/sketches-library/src/main/java/uk/gov/gchq/gaffer/sketches/datasketches/sampling/serialisation/ReservoirStringsUnionSerialiser.java [hunk 25]
```diff
@@ -53,4 +53,9 @@ public boolean canHandle(final Class clazz) {
     public boolean preservesObjectOrdering() {
         return false;
     }
+
+    @Override
+    public boolean isConsistent() {
+        return true;
```
Comment (p013570) [inline library/sketches-library/src/main/java/uk/gov/gchq/gaffer/sketches/datasketches/sampling/serialisation/ReservoirStringsUnionSerialiser.java]: This should probably be false.

File: library/sketches-library/src/main/java/uk/gov/gchq/gaffer/sketches/datasketches/theta/serialisation/SketchSerialiser.java [hunk 26]
```diff
@@ -51,4 +51,9 @@ public Sketch deserialiseEmpty() throws SerialisationException {
     public boolean preservesObjectOrdering() {
         return false;
     }
+
+    @Override
+    public boolean isConsistent() {
+        return true;
```
Comment (p013570) [inline library/sketches-library/src/main/java/uk/gov/gchq/gaffer/sketches/datasketches/theta/serialisation/SketchSerialiser.java]: This should probably be false.

File: library/sketches-library/src/main/java/uk/gov/gchq/gaffer/sketches/datasketches/theta/serialisation/UnionSerialiser.java [hunk 27]
```diff
@@ -54,4 +54,9 @@ public Union deserialiseEmpty() throws SerialisationException {
     public boolean preservesObjectOrdering() {
         return false;
     }
+
+    @Override
+    public boolean isConsistent() {
+        return true;
```
Comment (p013570) [inline library/sketches-library/src/main/java/uk/gov/gchq/gaffer/sketches/datasketches/theta/serialisation/UnionSerialiser.java]: This should probably be false.

File: library/spark/spark-accumulo-library/src/test/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/dataframe/MyPropertySerialiser.java [hunk 28]
```diff
@@ -54,4 +54,9 @@ public MyProperty deserialiseEmpty() {
     public boolean preservesObjectOrdering() {
         return false;
     }
+
+    @Override
+    public boolean isConsistent() {
+        return true;
```
Comment (p013570) [inline library/spark/spark-accumulo-library/src/test/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/handler/dataframe/MyPropertySerialiser.java]: This should probably be false.

File: library/time-library/src/main/java/uk/gov/gchq/gaffer/time/serialisation/BoundedTimestampSetSerialiser.java [hunk 29]
```diff
@@ -126,4 +126,9 @@ public BoundedTimestampSet deserialiseEmpty() throws SerialisationException {
     public boolean preservesObjectOrdering() {
         return false;
     }
+
+    @Override
+    public boolean isConsistent() {
+        return true;
```
Comment (p013570) [inline library/time-library/src/main/java/uk/gov/gchq/gaffer/time/serialisation/BoundedTimestampSetSerialiser.java]: This should probably be false.

File: library/time-library/src/main/java/uk/gov/gchq/gaffer/time/serialisation/RBMBackedTimestampSetSerialiser.java [hunk 30]
```diff
@@ -103,4 +103,9 @@ public RBMBackedTimestampSet deserialiseEmpty() throws SerialisationException {
     public boolean preservesObjectOrdering() {
         return false;
     }
+
+    @Override
+    public boolean isConsistent() {
+        return true;
```
Comment (p013570) [inline library/time-library/src/main/java/uk/gov/gchq/gaffer/time/serialisation/RBMBackedTimestampSetSerialiser.java]: This should be false.

File: store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/serialisation/impl/NestedHyperLogLogPlusParquetSerialiser.java [hunk 31]
```diff
@@ -77,6 +77,11 @@ public boolean preservesObjectOrdering() {
     }
 
     @Override
+    public boolean isConsistent() {
+        return true;
```
Comment (p013570) [inline store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/serialisation/impl/NestedHyperLogLogPlusParquetSerialiser.java]: This should probably be false.

**Comment by m607123** [comment]
Closed while tests fixed.

**Comment by m607123** [comment]
Closed while tests fixed.

**Pull request merged**